src/measure_theory/measure/finite_measure_weak_convergence.lean

```diff
@@ -382,6 +433,76 @@ def to_weak_dual_bounded_continuous_nnreal (μ : probability_measure α) :
   map_smul' := μ.to_finite_measure.test_against_nn_smul,
   cont := μ.test_against_nn_lipschitz.continuous, }
 
+lemma to_weak_dual_bounded_continuous_nnreal_eval_def
+  (μ : probability_measure α) (f : α →ᵇ ℝ≥0) :
+  μ.to_weak_dual_bounded_continuous_nnreal f = μ.test_against_nn f := rfl
+
+/-- The topology of weak convergence on `probability_measures α`. This is inherited (induced) from
+the weak-*  topology on `weak_dual ℝ≥0 (α →ᵇ ℝ≥0)` via the function
+`probability_measures.to_weak_dual_bcnn`. -/
+instance : topological_space (probability_measure α) :=
+topological_space.induced
+  (λ (μ : probability_measure α), μ.to_weak_dual_bounded_continuous_nnreal) infer_instance
```

Comment:
Is there a reason that you want the topology defined this way? I was expecting that it would be the topology induced by the coercion into `finite_measure`. Are these actually different? It seems like there is some code duplication that might be able to be avoided if you pass through the coercion instead.

Reply:
The subspace topology is the same as the topology defined here; in fact `probability_measure.to_finite_measure_embedding` proves it, I guess.

It is a good question which of the equivalent definitions of the topology is preferable in mathlib!

To probabilists, the topology of convergence in distribution means "the topology on Borel probability measures on a topological space, where convergence of probability measures amounts to the convergence of the expected values of each fixed bounded continuous random variable". The direct translation of that description is the induced topology from the pairing of probability measures with bounded continuous functions by integration. I chose this "textbook definition" as the definition of the topology on probability measures. (The only tweak was to consider just nonnegative bounded continuous functions, which you can achieve by adding constants. A subsequent PR will of course show the equivalence with the literal textbook definition...) But of course in mathlib we are not under any obligation to implement a notion in the textbook way, as long as the notion ends up being equivalent to what mathematicians mean by it and having the API that a mathematician would expect.

It is natural to define a similar topology for finite measures --- even some probabilistic results actually end up doing so. So I wrote both. I tend to think that the statements should exist for both (the majority of use cases will be probability measures, and the API should definitely bypass explicit coercions to finite measures most of the time). But code duplication in proofs should be avoided, as you say. Many proofs about probability measures already reduce to the corresponding result about finite measures. But there is some amount of repetition, certainly in statements, and also in proofs which were approximately one-liners for finite measures already.

If one were to define the topology on probability measures as the subspace topology from finite measures, then one could indeed avoid defining many things for probability measures, including the pairing of `probability_measure`s with test functions (bounded continuous non-negative). My feeling is that this test function pairing is nevertheless occasionally used. It would still exists via first coercing to a `finite_measure`, of course. But I think the canonical coercion from `probability_measure` should still be to just `measure` (the API about `measure` is much better than about `finite_measure`). Moreover, in that case some lemmas should still be proved about the subspace topology, I don't yet have a clear feeling for how large the code duplication would be.

Thanks for the comment! I don't yet have a clear opinion about this, although I lean towards the original induced topology option. I'd be happy to hear opinions of experienced formalizers and probabilists. And I will definitely think about it some more and probably test next weekend.

Reply:
I tested your suggestion (at least in some form). I think it is an improvement indeed.

This approach removed approximately 40 lines of (mostly duplicated) code. That included some lemmas that I now believe are not going to be necessary, as long as the API on `probability_measure` will help hide the fact that a coercion to `finite_measure` is involved in the definition of the topology. And indeed there is already a characterization of the `tendsto_nhds` along a general filter, whose content is basically the "textbook definition" of the topology of convergence in distribution, so I am relatively confident that hiding the implementation works sufficiently well. I took out the `test_against_nn` on probability measures as such, but kept `to_weak_dual_bcnn`, which is a "stronger version of the same".

Thank you for suggesting the improvement!